File: xcall/inref/ydbtypes.m
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;
; Copyright (c) 2024 YottaDB LLC and/or its subsidiaries.
; All rights reserved.
;
;	This source code contains the intellectual property
;	of its copyright holder(s), and is made available
;	under a license.  If you do not know the terms of
; the license, please stop and do not read further.
;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

run
 set int=-2147483647  ;-2^31-1: most negative int32 value
 set uint=4294967295  ;2^32-1: largest uint32 value
 set long=int
 set ulong=uint
 set int64=-9223372036854775807   ;-2^63: most negative int64 value
 set uint64=18446744073709551615  ;2^64-1: largest uint64 value
 set float=1234.56
 set double=17976931348623200000000000000000000000000000000  ;use full 15-digit double number accuracy
 set string1="a man"
 set string2="a plan"
 set string3="a canal"
 set string4="panama"
 set c=","

 write !,"### Loading call-out table"

 do &ydbtypes.noop()  ;Check whether call-out table loads

 write !,"### Test ydb_xxx_t types",!

 write "# Test input (I) parameters",!
 write "  Input:  ",int,c,int,c,uint,c,uint,c,long,c,long,c,ulong,c,ulong,c,int64,c,uint64,c,$fnumber(float,"",2),c,$fnumber(double/1E46,"",14)_"e+46",c,string1,c,string2,c,string3,c,string4,!
 do &ydbtypes.yinputs(int,int,uint,uint,long,long,ulong,ulong,int64,uint64,float,double,string1,string2,string3,string4)

 write "# Test output (O) parameters:",!
 set intO=1,uintO=2,longO=3,ulongO=4,int64O=5,uint64O=6,floatO=7,doubleO=8
 set string1O="test1",string2O="test2",string3O="test3",string4O="test4"
 write "  Input:  ",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!
 do &ydbtypes.youtputs(.intO,.uintO,.longO,.ulongO,.int64O,.uint64O,.floatO,.doubleO,.string1O,.string2O,.string3O,.string4O)
 write "  Receive:",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!

 write "# Test input/output (IO) parameters:",!
 set intO=int,uintO=uint,longO=long,ulongO=ulong,int64O=int64,uint64O=uint64,floatO=float,doubleO=double
 set string1O=string1,string2O=string2,string3O=string3,string4O=string4
 write "  Input:  ",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!
 do &ydbtypes.yio(.intO,.uintO,.longO,.ulongO,.int64O,.uint64O,.floatO,.doubleO,.string1O,.string2O,.string3O,.string4O)
 write "  Receive:",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!


 write !,"### Test standard C types",!

 write "# Test input (I) parameters",!
 write "  Input:  ",int,c,int,c,uint,c,uint,c,long,c,long,c,ulong,c,ulong,c,int64,c,uint64,c,$fnumber(float,"",2),c,$fnumber(double/1E46,"",14)_"e+46",c,string1,c,string2,c,string3,c,string4,!
 do &ydbtypes.cinputs(int,int,uint,uint,long,long,ulong,ulong,int64,uint64,float,double,string1,string2,string3,string4)

 write "# Test output (O) parameters:",!
 set intO=1,uintO=2,longO=3,ulongO=4,int64O=5,uint64O=6,floatO=7,doubleO=8
 set string1O="test1",string2O="test2",string3O="test3",string4O="test4"
 write "  Input:  ",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!
 do &ydbtypes.coutputs(.intO,.uintO,.longO,.ulongO,.int64O,.uint64O,.floatO,.doubleO,.string1O,.string2O,.string3O,.string4O)
 write "  Receive:",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!

 write "# Test input/output (IO) parameters:",!
 set intO=int,uintO=uint,longO=long,ulongO=ulong,int64O=int64,uint64O=uint64,floatO=float,doubleO=double
 set string1O=string1,string2O=string2,string3O=string3,string4O=string4
 write "  Input:  ",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!
 do &ydbtypes.cio(.intO,.uintO,.longO,.ulongO,.int64O,.uint64O,.floatO,.doubleO,.string1O,.string2O,.string3O,.string4O)
 write "  Receive:",intO,c,uintO,c,longO,c,ulongO,c,int64O,c,uint64O,c,$fnumber(floatO,"",2),c,$fnumber(doubleO/1E46,"",14)_"e+46",c,string1O,c,string2O,c,string3O,c,string4O,!

 ;gtm_platform_size is set by test system
 if $ztrnlnm("gtm_platform_size")'="32" do
 .write !,"### Test 64-bit ydb_xxx_t types",!
 .write "  Input:  ",int64,c,uint64,!
 .do &ydbtypes.yinputs64(int64,uint64)
 .write !,"### Test 64-bit standard C types",!
 .write "  Input:  ",int64,c,uint64,!
 .do &ydbtypes.cinputs64(int64,uint64)

 quit

File: xcall/inref/ydbtypes.c
/****************************************************************
*
* Copyright (c) 2024 YottaDB LLC and/or its subsidiaries.
* All rights reserved.
*
* This source code contains the intellectual property
* of its copyright holder(s), and is made available
* under a license.  If you do not know the terms of
* the license, please stop and do not read further.
*
****************************************************************/
/* Test various return values from external call-outs
 * Note: basic.csh / make_fcn.csh already test the following retvals:
 *   void, gtm_long_t, gtm_ulong_t, gtm_status_t, gtm_int_t, gtm_uint_t
 */
#include <limits.h>
#include <inttypes.h>
#include "gtm_stdio.h"
#include "gtmxc_types.h"

#define STRING1 "Do not be so open-minded"
#define STRING2 "that your brains"
#define STRING3 "fall out."
#define STRING4 "- G. K. Chesterton"

#define PRINT(...) PRINTF(__VA_ARGS__); fflush(stdout);

// Test standard types

// Dummy function used merely for trying invalid types in the .xc file
void noop(int _nargs) { }

void set_return_numbers(int*ip, uint*uip, long*lp, ulong*ulp, int64_t*i64p, uint64_t*ui64p, float*fp, double*dp) {
	*ip = -2133333333;
	*uip = 4188888888;
	*lp = -2133333333;
	*ulp = 4188888888;
	*i64p = -9111111111111111111LL;
	*ui64p = 17777777777777777777ULL;
	*fp = 1222.33;
	*dp = 1.66666666666666E+46;
}

void inputs_tester(int _nargs,
	int i, int*ip, uint ui, uint*uip, long l, long*lp, ulong ul, ulong*ulp, int64_t*i64p, uint64_t*ui64p,
	float*fp, double*dp, char*cp, ydb_string_t*sp, ydb_buffer_t*bp, ydb_char_t**cpp) {

	// Written in separate lines so that the test output will show exactly which parameter is at fault if there is a segfault
	PRINT("    Got:  %i,", i);
	PRINT("%i,", *ip);
	PRINT("%u,%u,", ui, *uip);
	PRINT("%li,%li,", l, *lp);
	PRINT("%lu,%lu,", ul, *ulp);
	PRINT("%" PRIi64 ",%" PRIu64 ",", *i64p, *ui64p);
	PRINT("%.2f,%.14e,", *fp, *dp);
	PRINT("%s,", cp);
	PRINT("%.*s,", (int)sp->length, sp->address);
	PRINT("%.*s,", (int)bp->len_used, bp->buf_addr);
	PRINT("%s\n", *cpp);
}

void outputs_tester(int _nargs,
	int*ip, uint*uip, long*lp, ulong*ulp, int64_t*i64p, uint64_t*ui64p,
	float*fp, double*dp, char*cp, ydb_string_t*sp, ydb_buffer_t*bp, char**cpp) {

	// Written in separate lines so that the test output will show exactly which parameter is at fault if there is a segfault
	PRINT("    Got:  %i,", *ip);
	PRINT("%u,", *uip);
	PRINT("%li,", *lp);
	PRINT("%lu,", *ulp);
	PRINT("%" PRIi64 ",", *i64p);
	PRINT("%" PRIu64 ",", *ui64p);
	PRINT("%.2f,", *fp);
	PRINT("%.14e,", *dp);
	PRINT("%s,", cp);
	PRINT("%.*s,", (int)sp->length, sp->address);
	PRINT("%.*s,", (int)bp->len_used, bp->buf_addr);
	PRINT("%s\n", *cpp);

	set_return_numbers(ip, uip, lp, ulp, i64p, ui64p, fp, dp);
	strcpy(cp, STRING1);
	sp->length = strlen(STRING2); strncpy(sp->address,  STRING2, sp->length);
	bp->len_used = strlen(STRING3); strncpy(bp->buf_addr, STRING3, bp->len_used);
	strcpy(*cpp, STRING4);

	PRINT("  Return: %i,%u,%li,%lu,%" PRIi64 ",%" PRIu64 ",%.2f,%.14e,%s,%.*s,%.*s,%s\n",
		*ip, *uip, *lp, *ulp, *i64p, *ui64p, *fp, *dp, cp, (int)sp->length, sp->address, (int)bp->len_used, bp->buf_addr, *cpp);
}

void io_tester(int _nargs,
	int*ip, uint*uip, long*lp, ulong*ulp, int64_t*i64p, uint64_t*ui64p,
	float*fp, double*dp, char*cp, ydb_string_t*sp, ydb_buffer_t*bp, char**cpp) {

	// Written in separate lines so that the test output will show exactly which parameter is at fault if there is a segfault
	PRINT("    Got:  %i,", *ip);
	PRINT("%u,", *uip);
	PRINT("%li,", *lp);
	PRINT("%lu,", *ulp);
	PRINT("%" PRIi64 ",", *i64p);
	PRINT("%" PRIu64 ",", *ui64p);
	PRINT("%.2f,", *fp);
	PRINT("%.14e,", *dp);
	PRINT("%s,", cp);
	PRINT("%.*s,", (int)sp->length, sp->address);
	PRINT("%.*s,", (int)bp->len_used, bp->buf_addr);
	PRINT("%s\n", *cpp);

	set_return_numbers(ip, uip, lp, ulp, i64p, ui64p, fp, dp);
	// output strings -- but can only output up to the length of the input string as pre-allocation is not allowed with IO
	strncpy(cp, STRING1, strlen(cp));
	sp->length = stpncpy(sp->address,  STRING2, sp->length) - sp->address;
	bp->len_used = stpncpy(bp->buf_addr, STRING3, bp->len_used) - bp->buf_addr;
	*stpncpy(*cpp, STRING4, strlen(*cpp)) = '\0';

	PRINT("  Return: %i,%u,%li,%lu,%" PRIi64 ",%" PRIu64 ",%.2f,%.14e,%s,%.*s,%.*s,%s\n",
		*ip, *uip, *lp, *ulp, *i64p, *ui64p, *fp, *dp, cp, (int)sp->length, sp->address, (int)bp->len_used, bp->buf_addr, *cpp);
}

void inputs64_tester(int _nargs, int64_t i64, uint64_t ui64) {
	PRINT("    Got:  %" PRIi64 ",", i64);
	PRINT("%" PRIu64 "\n", ui64);
}
